sstables_loader.cc

```diff
@@ -518,8 +554,15 @@ future<> sstables_loader::download_task_impl::run() {
     llog.debug("Streaming sstables from {}({}/{})", _endpoint, _bucket, _prefix);
     co_await _loader.invoke_on_all([this, &sstables_on_shards, table_id] (sstables_loader& loader) mutable -> future<> {
         co_await loader.load_and_stream(_ks, _cf, table_id, std::move(sstables_on_shards[this_shard_id()]), false, false,
-                                        [this] (unsigned num_streamed) {
-            _num_sstables_processed[this_shard_id()] += num_streamed;
+                                        [this] (double n) {
+            auto& processed = _batches_processed[this_shard_id()];
+            if (n > 0) {
+                // positive for delta
+                processed.first += n;
+            } else {
+                // negative for total
+                processed.second = -n;
+            }
```

Comment:
That's extremely confusing to use update value sign to indicate different pars of the progress being updated.
Can we replace `std::function<void(unsigned)> on_streamed` with `lw_shared_ptr<struct { unsigned progress; unsigned total; }> progress` and update its members in streamer methods directly?

Reply:
> That's extremely confusing to use update value sign to indicate different pars of the progress being updated. 

agreed.  neither am i in favor of this design. 

> Can we replace `std::function<void(unsigned)> on_streamed` with `lw_shared_ptr<struct { unsigned progress; unsigned total; }> progress` and update its members in streamer methods directly?

please note, i am using `function<void(double)>` in the latest revision. because in order to use a single interface of `sstable_streamer::stream_sstables()` shared by restore processes of both tables with tablets and vnodes, we cannot use "unsigned". because,  in `tablet_sstable_streamer::stream()`, we stream sstables on a per-tablet basis -- we iterate through the tablets' token ranges, and stream the sstables providing each tablet in batch size of a single tablet, and we need to advance the progress with a finer granularity of 16 sstables in each batch.

probably using the smart pointer is to be defensive and probably "safer", but i think, the coroutine or sub coroutine spawned by `loader.load_and_stream()` cannot outlive its caller. we can we just pass a raw pointer to the callee. 

but still, i don't think it's a good idea even we use a raw pointer, which exposes the internals of the caller to callee. so i will use `std::function<void(double, double)>`.

Reply:
>  i don't think it's a good idea even we use a raw pointer, which exposes the internals of the caller to callee. so i will use std::function<void(double, double)>.

I'm fine with using a callable to update progress, but the `std::function<void(double, double)>` is even more confusing :( How about introducing a `download_progress_monitor` class with two methods `.set_total()` (or `.update_total()` and `.update_progress()` (or `.add_progress()`)?

Reply:
hmm, it is confusing after rereading the code. i added two classes for tracking the progress, and accompanied them with some comment.